### PR TITLE
Dense vector field types updatable for int4

### PR DESCRIPTION
**Pull request by tteofili**
This PR makes it possible to update mappings also for the newly introduced int4 vectors (`int4_flat`, `int4_hnsw`), at the moment these are the allowed upgrade paths:
`flat` -> anything
`hnsw` -> `int8_hnsw`
`int8_flat` -> `hnsw`, `int8_hnsw`
`int8_hnsw` -> nothing
with int4 we may have the following upgrade paths:
`flat` -> anything
`hnsw` -> `int8_hnsw`, `int4_hnsw`
`int8_flat` -> `int4_flat`, `int8_hnsw`, `int4_hnsw`
`int4_flat` -> `hnsw`, `int8_hnsw`, `int4_hnsw`
`int8_hnsw` -> `int4_hnsw`
`int4_hnsw` -> nothing

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 1]
```diff
@@ -1562,15 +1574,17 @@ boolean supportsElementType(ElementType elementType) {
...
+            if (update.type.equals(this.type)) {
                 Int8HnswIndexOptions int8HnswIndexOptions = (Int8HnswIndexOptions) update;
                 // fewer connections would break assumptions on max number of connections (based on largest previous graph) during merge
                 // quantization could not behave as expected with different confidence intervals (and quantiles) to be created
                 updatable = int8HnswIndexOptions.m >= this.m;
                 updatable &= confidenceInterval == null
                     || int8HnswIndexOptions.confidenceInterval != null
                         && confidenceInterval.equals(int8HnswIndexOptions.confidenceInterval);
+            } else {
+                updatable = update.type.equals(VectorIndexType.INT4_HNSW.name);
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:1587]: We don't care about a lower number of connections when updating int8 to int4?

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:1587]: Agreed, the number of connections is important.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/180_update_dense_vector_type.yml [hunk 2]
```diff
@@ -1320,6 +1470,104 @@ setup:
...
                 confidence_interval: 0.3
 
+---
+"Allowed dense vector updates on same type but different other index_options, int4_hnsw":
+  - requires:
+      cluster_features: "gte_v8.16.0"
+      reason: 'updatable dense vector field type for int4 was added in 8.16'
+  - skip:
+      reason: "contains is a newly added assertion"
+      features: contains
```
Comment (mayya-sharipova) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/180_update_dense_vector_type.yml:1480]: I think the new way to speficy required runner feature is following:

```
  - requires:
      test_runner_features: [contains]
```

**Review by mayya-sharipova — APPROVED** [review]
@tteofili Thanks, overall LGTM! I've left a couple of comments.

Do we put somewhere in the documentation for users what types can be updated to what?

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java (hunk 1 above)
Comment (tteofili) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:1587]: very good point, thanks!

**Comment by tteofili** [comment]
thanks @benwtrent @mayya-sharipova for the reviews.
I think I've addressed your comments.

> Do we put somewhere in the documentation for users what types can be updated to what?
I am thinking to open a separate PR for the documentation bits.

**Pull request merged**